juju/apiclient.py

```diff
@@ -17,10 +25,250 @@
 logger = logging.getLogger("juju")
 
+
+class UnknownFacade(Exception):
+    """UnknownFacade is raised if the server doesn't have the requested facade.
+
+    This can occur if asking for a non-existent facade, or a newer facade that
+    is not supported on the API server, or even a known facade that isn't
+    supported at the endpoint that has been connected to.
+
+    """
+    def __init__(self, *args):
+        super(UnknownFacade, self).__init__(*args)
+
+
+class FacadeVersionNotSupported(Exception):
+    """FacadeVersionNotSupported is raised if the version requested isn't supported.
+
+    """
+    def __init__(self, *args):
+        super(UnknownFacade, self).__init__(*args)
+
+
+class ServerError(Exception):
+    """Base exception class for particular server side errors.
+
+    If there is no specific error type for the error code returned from the
+    server, then this is the type of the returned error.
+
+    """
+    def __init__(self, error_code, message):
+        self.error_code = error_code
+        self.message = message
+
+    def __str__(self):
+        return "<ServerError - '{}'>".format(message)
+
+
+class NotImplementedError(ServerError):
+    """NotImplementedError is raised when a method does not exist.
+
+    The error could be caused due to either the method name being incorred, or
+    the facade version not being supported on the API server, or the facade
+    name itself not existing on the server.
+
+    If the Facade instances are being constructed through the Connection
+    instance, then the more specialized UnknownFacade and
+    FacadeVersionNotSupported errors will be raised, however if a Facade
+    instance is created through another method, the NotImplementedError will
+    be raised in the cases of bad facade names or versions.
+
+    """
+    def __init__(self, *args):
+        super(NotImplementedError, self).__init__(*args)
+
+
+
+def new_error(response):
+    """Constructs a specific exception type based on the ErrorCode."""
+    err_code = response.get("ErrorCode")
+    message = response.get("Error")
+    if err_code == "not implemented":
+        return NotImplementedError(err_code, message)
+    return ServerError(err_code, message)
+
+
+class Facade(object):
+    """The Facade instances provide convenient syntactic sugar for API calls.
+
+    Facade instances are constructed with a Connection, and facade name and
+    version. Any resulting method calls to a facade are translated into RPC
+    calls on the connection instance.
+
+    For example, calling the 'FullStatus' method on the client facade:
+
+        >>> client = Facade(connection, "Client")
+        >>> result = client.FullStatus()
+
+    This ends up being the same as:
+
+        >>> connection._rpc("Client", "FullStatus")
+
+    """
+
+    def __init__(self, connection, name, version=None):
+        self.connection = connection
+        self.name = name
+        self.version = version
+
+    def __getattribute__(self, attr):
```

Comment:
If I understand this correctly, I believe this would be simpler as a __getattr__ special method, rather than the more generica __getattribute__. E.g.
```python
def __getattr__(self, attr):
    return functools.partial(self.connection._rpc, self.name, attr, version=self.version)
```
Does it work?